apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data.CommonTesting/RetryHelpers.cs

```diff
@@ -43,30 +54,95 @@ public static class RetryHelpers
+            var clock = SystemClock.Instance;
+            var scheduler = SystemScheduler.Instance;
+            var jitter = RetrySettings.RandomJitter;
+
+            DateTime start = DateTime.UtcNow;
+            DateTime end = start + s_timeout;
+            // Immediate initial retry, before the exponential delay starts.
+            TimeSpan retryDelay = TimeSpan.Zero;
+            while (true)
             {
-                // Try it once more...
-                return func();
+                try
+                {
+                    return func();
+                }
+                catch (SpannerException e) when (e.IsRetryable)
+                {
+                    TimeSpan actualDelay = jitter.GetDelay(retryDelay);
+                    DateTime expectedRetryTime = clock.GetCurrentDateTimeUtc() + actualDelay;
+                    if (expectedRetryTime > end)
+                    {
+                        throw;
+                    }
+                    scheduler.Sleep(actualDelay, CancellationToken.None);
+                    retryDelay = s_backoffSettings.NextDelay(retryDelay);
+                    Interlocked.Increment(ref _retries);
+                }
             }
         }
 
         private static async Task<T> ExecuteWithRetryAsync<T>(Func<Task<T>> func)
         {
-            try
+            Interlocked.Increment(ref _calls);
+
+            // Make it easy to move this into production code later on by using IClock/IScheduler/IJitter
+            var clock = SystemClock.Instance;
+            var scheduler = SystemScheduler.Instance;
+            var jitter = RetrySettings.RandomJitter;
+
+            DateTime start = DateTime.UtcNow;
+            DateTime end = start + s_timeout;
+            // Immediate initial retry, before the exponential delay starts.
+            TimeSpan retryDelay = TimeSpan.Zero;
+            while (true)
             {
-                return await func();
+                try
+                {
+                    return await func();
+                }
+                catch (SpannerException e) when (e.IsRetryable)
+                {
+                    TimeSpan actualDelay = jitter.GetDelay(retryDelay);
+                    DateTime expectedRetryTime = clock.GetCurrentDateTimeUtc() + actualDelay;
+                    if (expectedRetryTime > end)
+                    {
+                        throw;
+                    }
+                    await scheduler.Delay(actualDelay, CancellationToken.None).ConfigureAwait(false);
+                    retryDelay = s_backoffSettings.NextDelay(retryDelay);
+                    Interlocked.Increment(ref _retries);
+                }
             }
-            catch (SpannerException e) when (e.ErrorCode == ErrorCode.Aborted)
+        }
+
+        public static void MaybeLogStats(string description)
+        {
+            int calls = Interlocked.CompareExchange(ref _calls, 1, 1);
```

Comment:
nit: In both of these, can't you use Interlocked.Read instead? Seems clearer, unless I'm missing something.

Reply:
Not with an `int`. I could change the field types to `long`, but that would give the impression that we're expecting a huge number of calls.

Reply:
Oh weird there's not an overload for int, didn't realize.